Mathlib/RingTheory/Etale/Field.lean

```diff
@@ -0,0 +1,133 @@
+/-
+Copyright (c) 2024 Andrew Yang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Andrew Yang
+-/
+import Mathlib.RingTheory.Etale.Basic
+import Mathlib.RingTheory.Unramified.Field
+
+/-!
+# Etale algebras over fields
+
+## Main results
+
+Let `K` be a field, `A` be a `K`-algebra and `L` be a field extension of `K`.
+
+- `Algebra.FormallyEtale.of_isSeparable`:
+    If `L` is separable over `K`, then `L` is formally etale over `K`.
+- `Algebra.FormallyEtale.iff_isSeparable`:
+    If `L` is (essentially) of finite type over `K`, then `L/K` is etale iff `L/K` is separable.
+
+## References
+
+- [B. Iversen, *Generic Local Structure of the Morphisms in Commutative Algebra*][iversen]
+
+-/
+
+
+universe u
+
+variable (K A L : Type u) [Field K] [Field L] [CommRing A] [Algebra K A] [Algebra K L]
+
+open Algebra Polynomial
+
+open scoped TensorProduct
+
+namespace Algebra.FormallyEtale
+
+theorem of_isSeparable_aux [Algebra.IsSeparable K L] [EssFiniteType K L] :
+    FormallyEtale K L := by
+  have := FormallyUnramified.of_isSeparable K L
+  have := FormallyUnramified.finite_of_free (R := K) (S := L)
+  constructor
+  intros B _ _ I h
+  refine ⟨(FormallyUnramified.of_isSeparable K L).1 I h, ?_⟩
+  intro f
+  let pb := Field.powerBasisOfFiniteOfSeparable K L
+  obtain ⟨x, hx⟩ := Ideal.Quotient.mk_surjective (f pb.gen)
+  have helper : ∀ x, IsScalarTower.toAlgHom K B (B ⧸ I) x = Ideal.Quotient.mk I x := fun _ ↦ rfl
+  have hx' : Ideal.Quotient.mk I (aeval x (minpoly K pb.gen)) = 0 := by
+    rw [← helper, ← aeval_algHom_apply, helper, hx, aeval_algHom_apply, minpoly.aeval, map_zero]
+  obtain ⟨u, hu⟩ : ∃ u, (aeval x) (derivative (minpoly K pb.gen)) * u + 1 ∈ I := by
+    have := (isUnit_iff_ne_zero.mpr ((Algebra.IsSeparable.isSeparable K
+      pb.gen).aeval_derivative_ne_zero (minpoly.aeval K _))).map f
+    rw [← aeval_algHom_apply, ← hx, ← helper, aeval_algHom_apply, helper] at this
+    obtain ⟨u, hu⟩ := Ideal.Quotient.mk_surjective (-this.unit⁻¹ : B ⧸ I)
+    use u
+    rw [← Ideal.Quotient.eq_zero_iff_mem, map_add, map_mul, map_one, hu, mul_neg,
+      IsUnit.mul_val_inv, neg_add_cancel]
+  use pb.liftEquiv.symm ⟨x + u * aeval x (minpoly K pb.gen), ?_⟩
+  · apply pb.algHom_ext
+    simp [hx, hx']
+  · rw [← eval_map_algebraMap, Polynomial.eval_add_of_sq_eq_zero, derivative_map,
+      ← one_mul (eval x _), eval_map_algebraMap, eval_map_algebraMap, ← mul_assoc, ← add_mul,
+      ← Ideal.mem_bot, ← h, pow_two, add_comm]
+    exact Ideal.mul_mem_mul hu (Ideal.Quotient.eq_zero_iff_mem.mp hx')
```

Comment:
```suggestion
    · exact Ideal.mul_mem_mul hu (Ideal.Quotient.eq_zero_iff_mem.mp hx')
```
I thought we had a linter which would moan about this sort of thing now? But it didn't so I'll moan about it :-)